hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java

```diff
@@ -93,13 +94,11 @@
   protected static final String MINOR_COMPACTION_DROP_CACHE =
       "hbase.regionserver.minorcompaction.pagecache.drop";
 
-  private final boolean dropCacheMajor;
-  private final boolean dropCacheMinor;
+  protected final boolean dropCacheMajor;
+  protected final boolean dropCacheMinor;
 
-  // In compaction process only a single thread will access and write to this field, and
-  // getCompactionTargets is the only place we will access it other than the compaction thread, so
-  // make it volatile.
-  protected volatile T writer = null;
+  protected final Set<T> writers = new HashSet<>();
```

Comment:
writers = Collections.synchronizedSet(Collections.newSetFromMap(new IdentityHashMap<>()));

I think IdentityHashMap is safer as we do not know whether the implementation class for CellSink will implement equals or not. And also use synchronizedSet so we do not need to manually synchronize everywhere, the only place we need to manually synchronize is when we want to iterator over the it.

And the same to the CompactionProgress field.